Handle pruned columns in a pushed down pinot table scan

agrawaldevesh opened the pull request:
## Problem: _PruneUnreferencedOptimizer_ can prune columns after pushdown has been done.

_PruneUnreferencedOptimizer_ runs after the connector pushdown optimizer and prunes away columns it thinks are not used in the table scan node. This creates a mismatch between the generated PQL and the columns expected during actual scan. This can happen in both broker and segment pinot scan codepaths.

As an example, consider this query:
```
select count(*) from baseballstats group by teamid limit 10
```

In this query, the pinot connector sees the plan: 
```
Project (count) -> Limit 10 -> Aggregation (teamid, count(*)) -> Scan baseballstats.
```

Since projections cannot currently be pushed on top of aggregations, so only the plan until the limit is pushed down. The resulting plan looks like:
```
Project(count) -> Pushed table scan node with PQL: "select count(*) from baseballstats group by teamid".
```
Note that the pushed table scan will emit "teamid" and "count(*)". 

PruneUnreferencedOptimizer runs next and removes the "teamid" from the scan node. This creates a problem: we are emitting an extra column teamid in the PinotBrokerPageSource which the scan does not need. 

## Solution: Remember the original column handles that corresponded to the generated PQL

This PR attempts to handle this scenario: both in the broker and segment codepaths. It does so with a lot of plumbing: It remembers the original column handles when the PQL was generated in the _PinotTableHandle_. These are passed eventually to the page sources via the _PinotSplit_. 

The page source then checks which columns can be ignored or have been re-ordered. The broker page source already had some support for this (necessitated by the _hiddenColumnSet_ for handling group-by without aggregations) and I have merely generalized this support. It was also trivial to add this support to the segment page source.

Added unit tests for these conditions in the pinot broker and segment page sources.


```
== NO RELEASE NOTE ==
```

highker commented on presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotBrokerPageSource.java:
```diff
@@ -75,6 +78,7 @@
     private final GeneratedPql brokerPql;
     private final PinotConfig pinotConfig;
     private final List<PinotColumnHandle> columnHandles;
+    private List<PinotColumnHandle> expectedHandles;
```
final

highker commented on presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotBrokerPageSource.java:
```diff
@@ -169,14 +178,19 @@ private void setValuesForGroupby(
+        requireNonNull(values, "Expected valid values in group by");
+        Preconditions.checkState(
+                blockBuilders.size() == values.length + group.size(),
+                String.format(
```
keep all params for `format` in the same line or one param each line.

highker commented on presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotBrokerPageSource.java:
```diff
@@ -421,4 +430,60 @@ public void close()
         finished = true;
     }
+
+    static class BlockAndTypeBuilder
```
private

highker commented on presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotBrokerPageSource.java:
```diff
@@ -421,4 +430,60 @@ public void close()
+            columnBlockBuilders = new ArrayList<>();
+            columnTypes = new ArrayList<>();
+
+            for (int expectedColIdx : brokerPql.getExpectedColumnIndices()) {
```
`expectedColumnIndex`

highker commented on presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotBrokerPageSource.java:
```diff
@@ -421,4 +430,60 @@ public void close()
+            columnTypes = new ArrayList<>();
+
+            for (int expectedColIdx : brokerPql.getExpectedColumnIndices()) {
+                int actualColIdx = -1;
```
actualColumnIndex

highker commented on presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotBrokerPageSource.java:
```diff
@@ -421,4 +430,60 @@ public void close()
+                    .map(PinotColumnHandle::getDataType)
+                    .collect(Collectors.toList());
+            pageBuilder = new PageBuilder(expectedTypes);
+            Preconditions.checkState(brokerPql.getExpectedColumnIndices().size() == expectedColumnHandles.size());
```
static import `checkState`

highker commented on presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotSegmentPageSource.java:
```diff
@@ -85,6 +84,10 @@ public PinotSegmentPageSource(
         this.columnHandles = requireNonNull(columnHandles, "columnHandles is null");
         this.session = requireNonNull(session, "session is null");
+        this.columnTypes = columnHandles
+                .stream()
```
move `.stream()` to the previous line

highker replied:
not addressed?

highker commented on presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotSegmentPageSource.java:
```diff
@@ -99,6 +102,12 @@ private static void checkExceptions(DataTable dataTable, PinotSplit split)
         if (!exceptions.isEmpty()) {
             throw new PinotException(PinotErrorCode.PINOT_EXCEPTION, split.getSegmentPql(), String.format("Encountered %d pinot exceptions for split %s: %s", exceptions.size(), split, exceptions));
         }
+        int columnsExpected = split.getExpectedColumnHandles().size();
```
I would name them expectedColumns and actualColumns

highker replied:
not addressed?

highker commented on presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotSegmentPageSource.java:
```diff
@@ -99,6 +102,12 @@ private static void checkExceptions(DataTable dataTable, PinotSplit split)
+        int columnsExpected = split.getExpectedColumnHandles().size();
+        int columnsActual = dataTable.getDataSchema().size();
+        if (columnsActual != columnsExpected) {
+            throw new PinotException(PinotErrorCode.PINOT_EXCEPTION, split.getSegmentPql(), String.format(
```
one param per line for both `PinotException()` and `format()`

highker commented on presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotSegmentPageSource.java:
```diff
@@ -150,15 +159,22 @@ public Page getNextPage()
             // Write a block for each column in the original order.
-            writeBlock(blockBuilder, columnType, columnHandleIdx);
+            PinotColumnHandle handle = columnHandles.get(columnHandleIdx);
+            int idxReturnedByPinot = expectedColumnHandles.indexOf(handle);
```
index

highker replied:
not addressed?

highker commented on presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotSegmentPageSource.java:
```diff
@@ -150,15 +159,22 @@ public Page getNextPage()
+            int idxReturnedByPinot = expectedColumnHandles.indexOf(handle);
+            if (idxReturnedByPinot < 0) {
+                throw new PinotException(PinotErrorCode.PINOT_INVALID_PQL_GENERATED, split.getSegmentPql(),
+                        String.format("Expected column handle %s to be present in the handles %s corresponding to the segment PQL",
```
same

highker replied:
not addressed?

highker commented on presto-pinot-toolkit/src/test/java/com/facebook/presto/pinot/TestPinotBrokerPageSource.java:
```diff
@@ -161,26 +190,60 @@ PqlParsedInfo getBasicInfoFromPql(String pqlResponse)
+        PinotBrokerPageSource pageSource = new PinotBrokerPageSource(
+                pinotConfig,
+                new TestingConnectorSession(ImmutableList.of()),
+                generatedPql, actualHandles, expectedColumnHandles, new MockPinotClusterInfoFetcher(pinotConfig), objectMapper);
```
one param per line

highker commented on presto-pinot-toolkit/src/test/java/com/facebook/presto/pinot/TestPinotBrokerPageSource.java:
```diff
@@ -161,26 +190,60 @@ PqlParsedInfo getBasicInfoFromPql(String pqlResponse)
+            BlockBuilder builder = columnBlockBuilders.get(i);
+            assertEquals(type == null, builder == null);
+            if (type != null) {
+                ++numNonNullTypes;
```
numNonNullTypes++

highker commented on presto-pinot-toolkit/src/test/java/com/facebook/presto/pinot/TestPinotBrokerPageSource.java:
```diff
@@ -161,26 +190,60 @@ PqlParsedInfo getBasicInfoFromPql(String pqlResponse)
+        assertEquals(columnTypes.size(), columnBlockBuilders.size());
+
+        int numNonNullTypes = 0;
+        for (int i = 0; i < columnTypes.size(); ++i) {
```
i++

highker commented on presto-pinot-toolkit/src/test/java/com/facebook/presto/pinot/TestPinotBrokerPageSource.java:
```diff
@@ -190,16 +253,22 @@ public void testPopulateFromPql(String pql, String pqlResponse, List<Type> types
-        return new PinotBrokerPageSource(pinotConfig, new TestingConnectorSession(ImmutableList.of()), generatedPql, pinotColumnHandles, new MockPinotClusterInfoFetcher(pinotConfig), objectMapper);
+        int numValid = 0;
+        Set<Integer> uniqueIndices = new HashSet<>();
+        for (int i : expectedColumnIndices) {
```
`i` -> `expectedColumnIndex`

highker commented on presto-pinot-toolkit/src/test/java/com/facebook/presto/pinot/TestPinotBrokerPageSource.java:
```diff
@@ -190,16 +253,22 @@ public void testPopulateFromPql(String pql, String pqlResponse, List<Type> types
+        for (int i : expectedColumnIndices) {
+            assertTrue(i == -1 || i >= 0 && i < expectedColumnHandles.size());
+            if (i >= 0) {
+                ++numValid;
```
numValid++

highker commented on presto-pinot-toolkit/src/test/java/com/facebook/presto/pinot/TestPinotSegmentPageSource.java:
```diff
@@ -327,6 +327,47 @@ private DataTable createDataTableWithAllTypes()
+        PinotSplit mockPinotSplit = new PinotSplit(pinotConnectorId.toString(), PinotSplit.SplitType.SEGMENT, expectedColumnHandles, Optional.empty(), Optional.of("blah"), ImmutableList.of("seg"), Optional.of("host"));
+
+        ImmutableList.Builder<Integer> columnsSurvivingBuilder = ImmutableList.builder();
+        for (int i = expectedColumnHandles.size() - 1; i >= 0; --i) {
```
i--

highker commented on presto-pinot-toolkit/src/test/java/com/facebook/presto/pinot/TestPinotSegmentPageSource.java:
```diff
@@ -327,6 +327,47 @@ private DataTable createDataTableWithAllTypes()
+                int originalColumnIndex = columnsSurviving.get(j);
+                Type type = PinotColumnUtils.getPrestoTypeFromPinotType(getFieldSpec("dontcare", ALL_TYPES.get(originalColumnIndex)), false, false);
+                long maxHashCode = Long.MIN_VALUE;
+                for (int k = 0; k < NUM_ROWS; ++k) {
```
k++

highker commented on presto-pinot-toolkit/src/test/java/com/facebook/presto/pinot/TestPinotSplitManager.java:
```diff
@@ -61,8 +61,10 @@ private void testSegmentSplitsHelperNoFilter(PinotTableHandle table, int segment
-        PinotQueryGenerator.GeneratedPql generatedPql = new PinotQueryGenerator(pinotConfig, typeManager, functionMetadataManager, standardFunctionResolution).generate(plan, sessionHolder.getConnectorSession()).get().getGeneratedPql();
-        PinotTableHandle pinotTableHandle = new PinotTableHandle(table.getConnectorId(), table.getSchemaName(), table.getTableName(), Optional.of(false), Optional.of(generatedPql));
+        PinotQueryGenerator.PinotQueryGeneratorResult pinotQueryGeneratorResult = new PinotQueryGenerator(pinotConfig, typeManager, functionMetadataManager, standardFunctionResolution).generate(plan, sessionHolder.getConnectorSession()).get();
+        ImmutableList<PinotColumnHandle> expectedHandles = ImmutableList.copyOf(pinotQueryGeneratorResult.getContext().getAssignments().values());
```
List<PinotColumnHandle>

highker commented on presto-pinot-toolkit/src/test/java/com/facebook/presto/pinot/TestPinotSplitManager.java:
```diff
@@ -74,8 +76,10 @@ private void testSegmentSplitsHelperWithFilter(PinotTableHandle table, int segme
-        PinotQueryGenerator.GeneratedPql generatedPql = new PinotQueryGenerator(pinotConfig, typeManager, functionMetadataManager, standardFunctionResolution).generate(plan, sessionHolder.getConnectorSession()).get().getGeneratedPql();
-        PinotTableHandle pinotTableHandle = new PinotTableHandle(table.getConnectorId(), table.getSchemaName(), table.getTableName(), Optional.of(false), Optional.of(generatedPql));
+        PinotQueryGenerator.PinotQueryGeneratorResult pinotQueryGeneratorResult = new PinotQueryGenerator(pinotConfig, typeManager, functionMetadataManager, standardFunctionResolution).generate(plan, sessionHolder.getConnectorSession()).get();
+        ImmutableList<PinotColumnHandle> expectedHandles = ImmutableList.copyOf(pinotQueryGeneratorResult.getContext().getAssignments().values());
```
List<PinotColumnHandle>

highker commented:
@zhenxiao, could you help taking a look as well?

zhenxiao commented on presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotBrokerPageSource.java:
```diff
@@ -421,4 +432,61 @@ public void close()
+            Preconditions.checkState(expectedColumnHandles.size() >= columnHandles.size());
+
+            // map the expected column handle to the actual handle
+            int[] expectedToActual = new int[expectedColumnHandles.size()];
```
s/expectedToActual/handleMapping/g

zhenxiao commented on presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotBrokerPageSource.java:
```diff
@@ -421,4 +432,61 @@ public void close()
+            columnTypes = new ArrayList<>();
+
+            for (int expectedColumnIdx : brokerPql.getExpectedColumnIndices()) {
+                int actualColumnIdx = -1;
```
s/actualColumnIdx/columnIndex/g

zhenxiao commented on presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotBrokerPageSource.java:
```diff
@@ -421,4 +432,61 @@ public void close()
+            columnBlockBuilders = new ArrayList<>();
+            columnTypes = new ArrayList<>();
+
+            for (int expectedColumnIdx : brokerPql.getExpectedColumnIndices()) {
```
s/expectedColumnIdx/expectedColumnIndex/g

zhenxiao commented on presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotSegmentPageSource.java:
```diff
@@ -150,15 +159,22 @@ public Page getNextPage()
             Type columnType = columnTypes.get(columnHandleIdx);
             // Write a block for each column in the original order.
-            writeBlock(blockBuilder, columnType, columnHandleIdx);
+            PinotColumnHandle handle = columnHandles.get(columnHandleIdx);
```
s/columnHandleIdx/columnHandleIndex/g

zhenxiao commented on presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotSegmentPageSource.java:
```diff
@@ -150,15 +159,22 @@ public Page getNextPage()
+            PinotColumnHandle handle = columnHandles.get(columnHandleIdx);
+            int idxReturnedByPinot = expectedColumnHandles.indexOf(handle);
+            if (idxReturnedByPinot < 0) {
+                throw new PinotException(PinotErrorCode.PINOT_INVALID_PQL_GENERATED, split.getSegmentPql(),
```
static import PINOT_INVALID_PQL_GENERATED

zhenxiao commented on presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotSegmentPageSource.java:
```diff
@@ -255,10 +267,10 @@ else if (javaType.equals(Slice.class)) {
         else {
             throw new PrestoException(
-                    PINOT_UNSUPPORTED_COLUMN_TYPE,
+                    PinotErrorCode.PINOT_UNSUPPORTED_COLUMN_TYPE,
```
static import PINOT_UNSUPPORTED_COLUMN_TYPE

zhenxiao commented on presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/query/PinotQueryGeneratorContext.java:
```diff
@@ -358,7 +358,9 @@ else if (!groupByColumns.isEmpty()) {
 
-        List<Integer> indices = getIndicesMappingFromPinotSchemaToPrestoSchema(query, getAssignments());
+        LinkedHashMap<VariableReferenceExpression, PinotColumnHandle> assignments = getAssignments();
+        List<PinotColumnHandle> handles = ImmutableList.copyOf(assignments.values());
```
is handles used?

agrawaldevesh commented:
Thanks for the speedy reviews @highker and @zhenxiao !. The code reads better now :-).

highker commented on presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotBrokerPageSource.java:
```diff
@@ -169,14 +178,21 @@ private void setValuesForGroupby(
     {
+        requireNonNull(group, "Expected valid group");
+        requireNonNull(values, "Expected valid values in group by");
+        Preconditions.checkState(
```
static import checkState

highker commented on presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotBrokerPageSource.java:
```diff
@@ -274,6 +279,12 @@ private int issuePqlAndPopulate(
 
+    private static String asText(JsonNode node)
+    {
+        Preconditions.checkState(node.isValueNode());
```
static import

highker commented on presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotBrokerPageSource.java:
```diff
@@ -421,4 +432,69 @@ public void close()
+            List<Type> expectedTypes = columnHandles.stream()
+                    .map(PinotColumnHandle::getDataType)
+                    .collect(Collectors.toList());
+            pageBuilder = new PageBuilder(expectedTypes);
```
this.pageBuilder

highker commented on presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotBrokerPageSource.java:
```diff
@@ -421,4 +432,69 @@ public void close()
+                handleMapping[i] = columnHandles.indexOf(expectedColumnHandles.get(i));
+            }
+
+            columnBlockBuilders = new ArrayList<>();
```
this.columnBlockBuilders

highker commented on presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotBrokerPageSource.java:
```diff
@@ -421,4 +432,69 @@ public void close()
+            }
+
+            columnBlockBuilders = new ArrayList<>();
+            columnTypes = new ArrayList<>();
```
this

highker commented on presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotSegmentPageSource.java:
```diff
@@ -97,7 +105,13 @@ private static void checkExceptions(DataTable dataTable, PinotSplit split)
+        int columnsExpected = split.getExpectedColumnHandles().size();
+        int columnsActual = dataTable.getDataSchema().size();
+        if (columnsActual != columnsExpected) {
+            throw new PinotException(PINOT_EXCEPTION, split.getSegmentPql(), String.format(
```
one param per line

agrawaldevesh replied:
new change :-)

agrawaldevesh commented:
@highker for future reference, where can I find the style guide that you use. Perhaps we should add such nits to the style checker or encode them somewhere on a style guide? They are missing from https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines

agrawaldevesh commented:
@highker, thanks for calling me out to make the code be more consistent. I fixed all classes of nits including in the rest of the pinot connector code:

- Renamed idx to index everywhere 
- Used static imports for Preconditions and PinotErrorCode
- Moved params into one line each or all on the same line.

highker commented:
@agrawaldevesh, thanks for addressing the comments. Many nits are actually capture by https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting. For example, `.stream()` in the same line, one param per line, using immutable collections, or avoiding abbreviations, etc.

agrawaldevesh commented:
Thanks @highker. Now I know :)

This PullRequest is good to be merged in. The Travis failures are unrelated but someone should look into them since they signal some test infra issues.
